tools/audioconv64/conv_xm64.c

```diff
@@ -0,0 +1,208 @@
+/*
+ * We convert XM files to XM64 format. XM64 is a dump of the internal datastructure
+ * of libxm, similar to the official "libxmize" but portable across different
+ * architectures (and endian friendly).
+ *
+ * On top of this, XM64 has also several benefits and pre-processing:
+ *
+ *  * Samples with ping-pong loops are unrolled to forward loops, as the RSP
+ *    player does not support ping-pong looping.
+ *  * Patterns are recompressed using a custom RLE algorithm. This helps reducing
+ *    ROM size while still requiring negligible CPU time each time a new pattern
+ *    is loaded (only the current pattern is kept in RAM at any given time).
+ *    The decompression also requires no additional memory (RAM).
+ *  * The module is analyzed to calculate the minimum amount of RAM to be allocated
+ *    for each channel for streaming samples for ROMs. Each channel has a buffer
+ *    that must contain enough samples for playing one "tick", so the exact
+ *    size depends on the playing speed, sample pitch, etc. across the whole
+ *    module.
+ */
+
+#include "xm64internal.h"
+#include "mixer.h"
+
+// Bring libxm in
+#include "../../src/audio/libxm/play.c"
+#include "../../src/audio/libxm/context.c"
+#include "../../src/audio/libxm/load.c"
+
+int xm_convert(const char *infn, const char *outfn) {
+	if (flag_verbose)
+		fprintf(stderr, "Converting: %s => %s\n", infn, outfn);
+
+	FILE *xm = fopen(infn, "rb");
+	if (!xm) fatal("cannot open: %s\n", infn);
+
+	fseek(xm, 0, SEEK_END);
+	int fsize = ftell(xm);
+	fseek(xm, 0, SEEK_SET);
+
+	char *xmdata = malloc(fsize);
+	fread(xmdata, 1, fsize, xm);
+
+	size_t mem_ctx, mem_pat, mem_sam;
+	xm_get_memory_needed_for_context(xmdata, fsize, &mem_ctx, &mem_pat, &mem_sam);
+
+	xm_context_t* ctx;
+	xm_create_context_safe(&ctx, xmdata, fsize, 44100);
```

Comment:
Also would be nice to get it from the command arguments and default to 48k, WDYT?

Reply:
I rethought about this and I was confused. The sample buffer size does not depend on the output rate because resampling happens within the mixer. The sample buffer contains a waveform at its native playback rate. For instance, if the XM logic decides to play an instrument at 8000 Hz and the speed of XM is 25 ticks per second, for each tick I will need 8000 / 25 = 320 samples in the sample buffer, and the mixer will upscale those to 44100Hz or whatever output frequency is requested.

The number specified in the code does affect the internal calculations the XM player, but it doesn't change this end result. In fact, even changing this manually, there is no change in the calculated buffer size.

So there is no need for the command line option. I would just change this to 48000 just because it's a little less arbitrary than 44100Hz, and leave a comment that it doesn't affect the calculations anyway.